src/app/services/shell.service.ts

```diff
@@ -1,78 +1,104 @@
 import {
   Injectable, EventEmitter, Inject,
 } from '@angular/core';
+import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';
 import { environment } from 'environments/environment';
+import { Observable, tap } from 'rxjs';
+import { webSocket as rxjsWebsocket, WebSocketSubject } from 'rxjs/webSocket';
+import { IncomingApiMessageType } from 'app/enums/api-message-type.enum';
+import { WEBSOCKET } from 'app/helpers/websocket.helper';
 import { WINDOW } from 'app/helpers/window.helper';
 import { ShellConnectedEvent } from 'app/interfaces/shell.interface';
+import { TerminalConnectionData } from 'app/interfaces/terminal.interface';
 
+@UntilDestroy()
 @Injectable()
 export class ShellService {
-  pendingMessages: string[] = [];
-  socket: WebSocket;
-  connected = false;
-  token: string;
-  vmId: number;
-  podInfo: {
-    chart_release_name: string;
-    pod_name: string;
-    container_name: string;
-    command: string;
-  };
-
-  private shellCmdOutput: ArrayBuffer;
-  shellOutput = new EventEmitter<ArrayBuffer>();
-  shellConnected = new EventEmitter<ShellConnectedEvent>();
+  private encoder = new TextEncoder();
+  private ws$: WebSocketSubject<unknown>;
+  private connectionUrl = (this.window.location.protocol === 'https:' ? 'wss://' : 'ws://') + environment.remote + '/websocket/shell/';
+  private isConnected = false;
+
+  private token: string;
+  private connectionData: TerminalConnectionData;
+
+  private shellOutput = new EventEmitter<ArrayBuffer>();
+  private shellConnected = new EventEmitter<ShellConnectedEvent>();
+
+  get shellOutput$(): Observable<ArrayBuffer> {
+    return this.shellOutput.asObservable();
+  }
+
+  get shellConnected$(): Observable<ShellConnectedEvent> {
+    return this.shellConnected.asObservable();
+  }
 
   constructor(
     @Inject(WINDOW) private window: Window,
+    @Inject(WEBSOCKET) private webSocket: typeof rxjsWebsocket,
   ) {}
 
-  connect(): void {
-    this.socket = new WebSocket(
-      (this.window.location.protocol === 'https:' ? 'wss://' : 'ws://')
-        + environment.remote + '/websocket/shell/',
-    );
-    this.socket.onmessage = this.onmessage.bind(this);
-    this.socket.onopen = this.onopen.bind(this);
-    this.socket.onclose = this.onclose.bind(this);
+  connect(connectionData: TerminalConnectionData, token?: string): void {
+    this.connectionData = connectionData;
```

Comment:
`token` and `connectionData` don't need to be class members here. We can use the `bind` method to pass the params to `onOpen` and remove the class variables. The calling component can hold the token value of token in a variable to pass it on when reconnecting.